service/server/router_linux.cpp

```diff
@@ -170,3 +175,122 @@ void RouterLinux::flushDns()
     else
         qDebug().noquote() << "OUTPUT systemctl restart nscd/systemd-resolved: " + output;
 }
+
+bool RouterLinux::createTun(const QString &dev, const QString &subnet) {
+    qDebug().noquote() << "createTun start";
+
+    char cmd [1000] = {0x0};
+    sprintf(cmd, "ip tuntap add mode tun dev %s", dev.toStdString().c_str());
+    int sys = system(cmd);
+    if(sys < 0)
+    {
+        qDebug().noquote() << "Could not activate tun device!\n";
+        return false;
+    }
+    memset(&cmd, 0, sizeof(cmd));
+    sprintf(cmd, "ip addr add %s/24 dev %s", subnet.toStdString().c_str(), dev.toStdString().c_str());
+    sys = system(cmd);
+    if(sys < 0)
+    {
+        qDebug().noquote() << "Could not activate tun device!\n";
+        return false;
+    }
+    memset(&cmd, 0, sizeof(cmd));
+    sprintf(cmd, "ip link set dev %s up", dev.toStdString().c_str());
+    sys = system(cmd);
+    if(sys < 0)
+    {
+        qDebug().noquote() << "Could not activate tun device!\n";
+        return false;
+    }
+
+    return true;
+}
+
+bool RouterLinux::deleteTun(const QString &dev)
```

Comment:
may be just use `ip link delete <device>` ?

Reply:
We use system functions for tun device deletion. That is safer than shell command execution.